src/NServiceBus.Transport.RabbitMQ/Receiving/MessageConverter.cs

```diff
@@ -39,7 +39,12 @@ public string RetrieveMessageId(BasicDeliverEventArgs message, Dictionary<string
                 messageHeaders.Remove(BasicPropertiesExtensions.ConfirmationIdHeader);
             }
 
-            var deserializedHeaders = DeserializeHeaders(messageHeaders);
+            // Leaving space for ReplyTo, CorrelationId, DeliveryMode, EnclosedMessageTypes and CallbackQueue conditionally
+            // added below. This is a bit cumbersome and need to be changed when things are conditionally added below
+            // but it prevents the header dictionary from growing and relocating which creates quite a bit of
+            // memory allocations and eats up CPU cycles.
+            const int extraCapacity = 5;
```

Comment:
I see that you've counted 5 things here, but there are actually only 4 possible additional things that get added below.

`CallbackQueue` gets looked for, but it's another `ReplyTo` potentially added in that case.

Reply:
More generally, I'm wondering how important this extra capacity actually is. In most normal cases, I'm not sure these would even be adding anything to the header dictionary.

The more impactful change I suspect is setting the initial capacity at all, using the size of the `IDictionary`.

Reply:
It is important because when you set the capacity and then go over even with one it starts growing again because you asked for an initial capacity. This change is about setting a reasonable capacity that fits all cases and we don't have to grow the capacity. I guess I have to learn counting then :D *back to primary school Daniel!*